File: apps/menubar/Sources/ShipkitKit/Protocol.swift
import Foundation

public let protocolVersion = 1

public struct Warning: Codable, Equatable, Sendable {
    public let check: String
    public let message: String

    public init(check: String, message: String) {
        self.check = check
        self.message = message
    }
}

/// One request per connection, exactly as `src/approval/protocol.ts` emits it.
public struct ApprovalRequest: Codable, Equatable, Sendable {
    public let protocolVersion: Int
    public let fingerprint: String
    public let repo: String
    public let branch: String
    public let base: String
    public let head: String
    public let title: String
    public let commitMessage: String
    public let diffstat: String
    public let warnings: [Warning]

    // `protocol` is a Swift keyword. Without this mapping the decoder quietly
    // finds no version at all and every request looks like a version mismatch.
    enum CodingKeys: String, CodingKey {
        case protocolVersion = "protocol"
        case fingerprint, repo, branch, base, head, title, commitMessage, diffstat, warnings
    }
}

public enum Decision: String, Codable, Sendable {
    case approved, denied, pending
}

public struct ApprovalResponse: Codable, Equatable, Sendable {
    public let protocolVersion: Int
    public let fingerprint: String
    public let decision: Decision

    public init(protocolVersion: Int, fingerprint: String, decision: Decision) {
        self.protocolVersion = protocolVersion
        self.fingerprint = fingerprint
        self.decision = decision
    }

    enum CodingKeys: String, CodingKey {
        case protocolVersion = "protocol"
        case fingerprint, decision
    }
}

public func decodeRequest(_ line: Data) throws -> ApprovalRequest {
    try JSONDecoder().decode(ApprovalRequest.self, from: line)
}

public func encodeResponse(_ response: ApprovalResponse) throws -> Data {
    var data = try JSONEncoder().encode(response)
    data.append(0x0A)
    return data
}

/// One item shipkit has to say about the change, as the panel draws it.
public struct OfferedItem: Codable, Equatable, Sendable, Identifiable {
    /// `finding`, `warning`, `readiness` or `advice` — which channel it came out of.
    public let kind: String
    /// The check id, rule id or topic, whichever the channel spells it as.
    public let id: String
    public let message: String
    /// `refuses`, `warns` or `informs`. How hard shipkit is pushing.
    public let severity: String

    public init(kind: String, id: String, message: String, severity: String) {
        self.kind = kind
        self.id = id
        self.message = message
        self.severity = severity
    }
}

/// One file the change touches, and where to put the cursor when it opens.
public struct OfferedFile: Codable, Equatable, Sendable, Identifiable {
    public let path: String
    public let status: String
    public let line: Int

    /// The path is the identity: `readPushDiff` never returns a path twice.
    public var id: String { path }

    public init(path: String, status: String, line: Int) {
        self.path = path
        self.status = status
        self.line = line
    }
}

/// A review offered while `shipkit review` is also serving its page.
///
/// `kind` is what tells this apart from an approval on the wire: an approval
/// request carries no `kind`, a Keychain token request carries `kind: "token"`.
/// The convention was already here, so a third kind needed no version bump.
///
/// The diff is deliberately absent. A popover is the wrong shape for four
/// hundred files, the page is one click away, and `files` carries what the
/// panel actually needs — a row per file with a button that opens it.
public struct ReviewRequest: Codable, Equatable, Sendable {
    public let protocolVersion: Int
    public let kind: String
    public let fingerprint: String
    /// The repository's name, for the panel to display.
    public let repo: String
    /// The checkout's absolute path. The editor button resolves a file against it.
    public let root: String
    public let branch: String
    public let base: String
    /// Empty when the review is running without an agent's answer.
    public let commitMessage: String
    public let diffstat: String
    public let items: [OfferedItem]
    public let files: [OfferedFile]

    enum CodingKeys: String, CodingKey {
        case protocolVersion = "protocol"
        case kind, fingerprint, repo, root, branch, base, commitMessage, diffstat, items, files
    }

    public init(
        protocolVersion: Int, kind: String, fingerprint: String,
        repo: String, root: String, branch: String, base: String,
        commitMessage: String, diffstat: String,
        items: [OfferedItem], files: [OfferedFile]
    ) {
        self.protocolVersion = protocolVersion
        self.kind = kind
        self.fingerprint = fingerprint
        self.repo = repo
        self.root = root
        self.branch = branch
        self.base = base
        self.commitMessage = commitMessage
        self.diffstat = diffstat
        self.items = items
        self.files = files
    }
}

/// One thing a person ticked, with whatever they wrote about it.
public struct SelectedItem: Codable, Equatable, Sendable {
    public let kind: String
    public let id: String
    public let message: String
    /// What they want done. Empty when they ticked the box and said nothing.
    public let note: String

    public init(kind: String, id: String, message: String, note: String) {
        self.kind = kind
        self.id = id
        self.message = message
        self.note = note
    }
}

/// What the person said about a review.
///
/// Two answers rather than one with an empty list, because the difference is
/// what the command does next: an empty selection clears whatever was pending,
/// and that is a decision, not the absence of one.
public enum ReviewAnswer: String, Codable, Sendable {
    case selected, nothing
}

public struct ReviewResponse: Codable, Equatable, Sendable {
    public let protocolVersion: Int
    public let kind = "review"
    public let fingerprint: String
    public let answer: ReviewAnswer
    public let items: [SelectedItem]

    public init(protocolVersion: Int, fingerprint: String, answer: ReviewAnswer, items: [SelectedItem]) {
        self.protocolVersion = protocolVersion
        self.fingerprint = fingerprint
        self.answer = answer
        self.items = items
    }

    enum CodingKeys: String, CodingKey {
        case protocolVersion = "protocol"
        case kind, fingerprint, answer, items
    }
}

public func decodeReviewRequest(_ line: Data) throws -> ReviewRequest {
    try JSONDecoder().decode(ReviewRequest.self, from: line)
}

public func encodeReviewResponse(_ response: ReviewResponse) throws -> Data {
    var data = try JSONEncoder().encode(response)
    data.append(0x0A)
    return data
}

File: src/approval/protocol.ts
import { z } from "zod";
import type { Warning } from "../preflight/types.js";

export const PROTOCOL_VERSION = 1;

export class ProtocolError extends Error {}

export type ApprovalRequest = {
  protocol: number;
  fingerprint: string;
  repo: string;
  branch: string;
  base: string;
  head: string;
  title: string;
  commitMessage: string;
  diffstat: string;
  warnings: Warning[];
};

const responseSchema = z.object({
  protocol: z.number(),
  fingerprint: z.string().min(1),
  decision: z.enum(["approved", "denied", "pending"]),
});

export type ApprovalResponse = z.infer<typeof responseSchema>;

/** One request per connection, on one line. JSON escapes any newline a commit message carries. */
export function encodeRequest(request: ApprovalRequest): string {
  return `${JSON.stringify(request)}\n`;
}

export function decodeResponse(line: string, expectedFingerprint: string): ApprovalResponse {
  let parsed: unknown;
  try {
    parsed = JSON.parse(line);
  } catch (error) {
    const detail = error instanceof Error ? error.message : String(error);
    throw new ProtocolError(`Cannot parse the approval response: ${detail}`);
  }

  const result = responseSchema.safeParse(parsed);
  if (!result.success) {
    const detail = result.error.issues
      .map((issue) => `${issue.path.join(".")}: ${issue.message}`)
      .join("; ");
    throw new ProtocolError(`Malformed approval response: ${detail}`);
  }

  // Version before fingerprint: a version disagreement explains a fingerprint
  // mismatch, and reporting the symptom would send the reader the wrong way.
  if (result.data.protocol !== PROTOCOL_VERSION) {
    throw new ProtocolError(
      `The approval surface speaks protocol ${result.data.protocol}; this shipkit speaks ${PROTOCOL_VERSION}`,
    );
  }

  // A reply about some other request is not an answer to this one.
  if (result.data.fingerprint !== expectedFingerprint) {
    throw new ProtocolError("The approval response is for a different request");
  }

  return result.data;
}

/**
 * A review offered to the menu bar while `shipkit review` is also serving its page.
 *
 * `kind` is what tells this apart from an approval on the wire. The convention is already
 * there and already tested: an approval request carries no `kind`, a Keychain token request
 * carries `kind: "token"`. So a third kind needs no version bump and no compatibility story.
 *
 * The diff is deliberately not here. A popover is the wrong shape for four hundred files,
 * the page is one click away, and `files` carries what the panel actually needs — a row per
 * file with a button that opens it where a person can read it properly.
 */
export type OfferedItem = {
  kind: string;
  id: string;
  message: string;
  severity: "refuses" | "warns" | "informs";
};

export type OfferedFile = {
  path: string;
  status: string;
  /** Where to put the cursor when the editor opens. 1 when nothing better is known. */
  line: number;
};

export type ReviewOffer = {
  protocol: number;
  kind: "review";
  fingerprint: string;
  repo: string;
  /** The checkout's absolute path. The panel's editor button resolves against it. */
  root: string;
  /** shipkit's own review page, token and all. The panel's primary action opens it. */
  url: string;
  branch: string;
  base: string;
  /** Empty when the review is running without an agent's answer. */
  commitMessage: string;
  diffstat: string;
  items: OfferedItem[];
  files: OfferedFile[];
};

const selectedItemSchema = z.object({
  // The same five kinds `FixRequestKind` has. `comment` cannot come from the panel today —
  // the offer carries no diff for anyone to write on — but the two lists are one contract,
  // and letting them drift would mean a selection this shipkit writes is one it could not
  // then read back.
  kind: z.enum(["warning", "finding", "readiness", "advice", "comment"]),
  id: z.string().min(1),
  message: z.string(),
  note: z.string(),
});

const reviewResponseSchema = z.object({
  protocol: z.number(),
  kind: z.literal("review"),
  fingerprint: z.string().min(1),
  /**
   * `selected` carries items; `nothing` is a person saying they read it and it is fine.
   *
   * Two answers rather than one with an empty list, because the difference is what the
   * command does next: an empty selection clears whatever was pending, and that is a
   * decision, not the absence of one.
   */
  answer: z.enum(["selected", "nothing"]),
  items: z.array(selectedItemSchema),
});

export type ReviewResponse = z.infer<typeof reviewResponseSchema>;

export function encodeReviewOffer(offer: ReviewOffer): string {
  return `${JSON.stringify(offer)}\n`;
}

export function decodeReviewResponse(line: string, expectedFingerprint: string): ReviewResponse {
  let parsed: unknown;
  try {
    parsed = JSON.parse(line);
  } catch (error) {
    const detail = error instanceof Error ? error.message : String(error);
    throw new ProtocolError(`Cannot parse the review response: ${detail}`);
  }

  const result = reviewResponseSchema.safeParse(parsed);
  if (!result.success) {
    const detail = result.error.issues
      .map((issue) => `${issue.path.join(".")}: ${issue.message}`)
      .join("; ");
    throw new ProtocolError(`Malformed review response: ${detail}`);
  }

  // Version before fingerprint, for the reason `decodeResponse` gives: a version
  // disagreement explains a fingerprint mismatch, and reporting the symptom would send the
  // reader the wrong way.
  if (result.data.protocol !== PROTOCOL_VERSION) {
    throw new ProtocolError(
      `The approval surface speaks protocol ${result.data.protocol}; this shipkit speaks ${PROTOCOL_VERSION}`,
    );
  }

  if (result.data.fingerprint !== expectedFingerprint) {
    throw new ProtocolError("The review response is for a different review");
  }

  // An answer of `nothing` that carries items is a surface contradicting itself, and the two
  // readings lead opposite ways — write these, or clear what is pending. Refused rather than
  // guessed.
  if (result.data.answer === "nothing" && result.data.items.length > 0) {
    throw new ProtocolError("The review response says nothing was selected but carries items");
  }

  return result.data;
}
